File: cpp/opencv/yolo_opencv_semantic.cpp
/* 
 * @Author: taifyang
 * @Date: 2026-01-03 20:51:36
 * @LastEditTime: 2026-08-30 11:26:30
 * @Description: source file for YOLO opencv semantic segmentation
 */

 #include "yolo_opencv.h"

 void YOLO_OpenCV_Semantic::init(const Algo_Type algo_type, const Device_Type device_type, const Model_Type model_type, const std::string model_path)
{
	if (algo_type != YOLO26)
	{
		std::cerr << "unsupported algo type!" << std::endl;
		std::exit(-1);
	}
	YOLO_OpenCV::init(algo_type, device_type, model_type, model_path);
	YOLO_Semantic::init(algo_type, device_type, model_type, model_path);
}

void YOLO_OpenCV_Semantic::pre_process()
{
	YOLO_OpenCV_Detect::pre_process();
}

void YOLO_OpenCV_Semantic::post_process()
{
	m_output0_host = (uint8_t*)m_output[0].data;

	m_mask = cv::Mat::zeros(m_image.size(), CV_8UC1);
	cv::Mat output = cv::Mat::zeros(m_input_size, CV_8UC1);
	std::copy(m_output0_host, m_output0_host + m_output_numdet, output.data);
	scale_mask(output, m_mask, m_input_size, m_image.size());

	if (m_draw_result)
	{
		draw_result();
	}
}

File: python/backends/OpenCV/yolo_opencv_semantic.py
'''
Author: taifyang  
Date: 2024-06-12 22:23:07
LastEditTime: 2026-09-11 21:49:02
Description: opencv inference class for YOLO semantic segmentation algorithm
'''


from backends.utils import *
from backends.yolo_semantic import *
from backends.OpenCV.yolo_opencv import *


'''
description: opencv inference class for the YOLO semantic segmentation algorithm
'''    
class YOLO_OpenCV_Semantic(YOLO_OpenCV, YOLO_Semantic):
    '''
    description:            construction method
    param {*} self          instance of class
    param {str} algo_type   algorithm type
    param {str} device_type device type
    param {str} model_type  model type
    param {str} model_path  model path
    return {*}
    '''   
    def __init__(self, algo_type:str, device_type:str, model_type:str, model_path:str) -> None:
        YOLO_OpenCV.__init__(self, algo_type, device_type, model_type, model_path)
        YOLO_Semantic.__init__(self, algo_type, device_type, model_type, model_path)

    '''
    description:    model pre-process
    param {*} self  instance of class
    return {*}
    ''' 
    def pre_process(self) -> None:
        input = letterbox(self.image, self.inputs_shape)
        self.inputs = cv2.dnn.blobFromImage(input, 1/255., size=self.inputs_shape, swapRB=True, crop=False)
        self.net.setInput(self.inputs)
        
    '''
    description:    model post-process
    param {*} self  instance of class
    return {*}
    '''           
    def post_process(self) -> None:
        output = np.squeeze(self.outputs[0]).astype(dtype=np.float32)
        class_map = scale_masks(output, self.inputs_shape, self.image.shape)
        mask = class_map.astype(np.uint8)
        if self.render_result:
            masks = []
            for cls_id in np.unique(mask):
                binary_mask = (mask == cls_id).astype(np.bool_)
                masks.append(binary_mask)
            self.result = self.draw_result(masks)